Bulk queuing (#145)
andycroll/andycroll.com: _posts/2022/2022-01-17-enqueue-jobs-quickly-with-sidekiq-bulk.md

---

> title: "Use Sidekiq Directly, Not Through Active Job"

Author: ```suggestion
title: "Enqueue Jobs Quickly with Sidekiq’s Bulk Features"
```

> description: "Active Job is great, but you might be missing out."

Author: ```suggestion
description: "From thousands of network calls, to one."
```

layout: article
category: ruby
image:
  base: '2022/enqueue-jobs-quickly-with-sidekiq-bulk'
  alt: "Martial Arts on the beach"
  credit: "Thao Le Hoang"
  source: "https://unsplash.com/photos/Xl-ilWBKJNk"

---

> We often have to enqueue lots of the same jobs at the same time. Perhaps a data migration for all of one kind of object? Or a data processing step that will run faster if it's parallelised?

Reviewer: *lots of the same type of job (?)

Author: ```suggestion
We often have to enqueue lots of the same type of job at the same time. Perhaps a data migration for all of one kind of object? Or a data processing step that will run faster if it's parallelised?
```

> Our focus is often on the potentially large queues or the elapsed time it takes to churn through all the work. There is another task that we have to consider. Enqueing all that work in the first place.

Reviewer: *There is another part of this that we have to consider (?)

Reviewer: And then a colon, instead of a full-stop. (consider: *enqueuing all of that work in the first place.)

Author: ```suggestion
Our focus is often on the potentially large queues or the elapsed time it takes to churn through all the work. There is another part of this that we have to consider: enqueuing all that work in the first place.
```

## Instead of…

…looping through lots of Active Record objects:

```ruby
Customer.all.each do |customer|
  DeliverEmailJob.perform_async(customer)
end
```

...or better, if you have more than a few hundred jobs to enqueue:

> 

Reviewer: *or better, using `find_in_batches`, if you have...

> ...or better, if you have more than a few hundred jobs to enqueue:

Author: ```suggestion
...or better, using `find_in_batches`, if you have more than a few hundred jobs to enqueue:
```

```ruby
Customer.find_in_batches do |customer|
  DeliverEmailJob.perform_async(customer)
end
```

## Use…

…Sidekiq directly:

```ruby
ids = Customer.all.pluck(:id)

> array_of_args = ids.map { |x| [x] }

Reviewer: Does this work if you have millions of customers?

> DeliverEmailJob.perform_bulk(array_of_args)

Reviewer: I should have been using this! 😬

Author: Genuinely a game changer.

> ids = Customer.all.pluck(:id)
> array_of_args = ids.map { |x| [x] }
> DeliverEmailJob.perform_bulk(array_of_args)

Author: ```suggestion
Customer.all.in_batches do |relation|
  array_of_args = relation.ids.map { |x| [x] }
  DeliverEmailJob.perform_bulk(array_of_args)
end
```

```

Or, if using a Sidekiq version before 6.3.0:

```ruby

> ids = Customer.all.pluck(:id)
> array_of_args = ids.map { |x| [x] }
> Sidekiq::Client.push_bulk('class' => DeliverEmailJob, 'args' => array_of_args)

Author: ```suggestion
Customer.all.in_batches do |relation|
  array_of_args = relation.ids.map { |x| [x] }
  Sidekiq::Client.push_bulk('class' => DeliverEmailJob, 'args' => array_of_args)
end
```

```

## Why?

> Primarily this minimizes round trips to Redis. Instead of making an individual call for each enqueing action across the network, you make _one_. This is a big deal even when using code as well optimised for speed as Redis.

Reviewer: *This minimises (are we American now?) round trips to Redis...

Reviewer: *even when using _tools_(?) as well optimised for speed as Redis

Author: ```suggestion
This minimises round trips to Redis. Instead of making an individual call for each enqueing action across the network, you make _one_. This is a big deal even when using tools as well optimised for speed as Redis.
```

> This can also have the benefit (on top of passing scaler `id`s to jobs) of minimising the amount of memory used. By `pluck`ing `id`s rather than looping over arrays of Active Record models you are simply dealing with less complex objects which take up less memory.

Reviewer: Comma before 'which'

Reviewer: Is the part in parentheses needed? Isn't the additional benefit you're describe there because you are passing scalar IDs?

Author: ```suggestion
This can also have the benefit of minimising the amount of memory used. By `pluck`ing `id`s, or using the `ids` method as shown, rather than looping over arrays of Active Record models you are simply dealing with less complex objects which take up less memory.
```

> The author of Sidekiq [recommends limiting to 1000 jobs](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing) per bulk enqueue, that's the default in the `perform_bulk` method. Even then you are saving the time of 999 round trips to Redis.

Reviewer: *recommends a limit of 1000 jobs per bulk enqueue. That's the default in...

Author: ```suggestion
The author of Sidekiq [recommends a limit of 1,000 jobs](https://github.com/mperham/sidekiq/wiki/Bulk-Queueing) per bulk enqueue, that's the default in the `perform_bulk` method. Even then you are saving the time of 999 round trips to Redis.
```

## Why not?

> Typically the tasks that require many concurrent jobs are not enqueued during a web request from a customer, so you're typically doing this in a way that doesn't impact the response times of your customers directly.

Reviewer: This paragraph is a little unclear.

Author: ```suggestion
It is more usual that enqueuing lots of jobs does not occur during a web request from a customer. Thus your customers are not necessarily suffering through the much slower approach. However if you are enqueuing jobs on the command line or in scheduled tasks it’s now _your time_ you are wasting!
```

> The pre-6.3.0 version of the bulk API is a little fiddly and leaves space for you to make a mess of the method. I've done that a few times.

Reviewer: Better to explain how or remove this!

Author: ```suggestion
The pre-6.3.0 version of the bulk API is a little fiddly and leaves space for you to make a mess of the method. I'd recommend updating to the latest version and using the `.perform_bulk` syntax.
```

> This functionality requires that you use Sidekiq directly and not via Active Job, but there are [plenty of reasons to use Sidekiq like that](/ruby/use-sidekiq-directly-not-through-active-job/).

Reviewer: Again, could be better worded.

And maybe a form of 'also' since it's the 3rd point in this section.

Author: ```suggestion
Also, this functionality requires that you use Sidekiq directly and not via Active Job. I have previously suggested [plenty of reasons to use Sidekiq directly](/ruby/use-sidekiq-directly-not-through-active-job/).
```
